core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/Envelope.kt

```diff
@@ -0,0 +1,384 @@
+import fr.sncf.osrd.reporting.exceptions.ErrorType
+import fr.sncf.osrd.reporting.exceptions.OSRDError
+import java.util.*
+import java.util.stream.Stream
+
+@SuppressFBWarnings("URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD")
+class Envelope private constructor(parts: Array<EnvelopePart>) :
+    Iterable<EnvelopePart>, SearchableEnvelope, EnvelopeInterpolate {
+    private val parts: Array<EnvelopePart>
+    @JvmField val continuous: Boolean
+
+    // region CACHE FIELDS
+    /** Contains the position of all transitions, including the beginning and end positions */
+    private val partPositions: DoubleArray
+
+    // these two fields could be public, but aren't for the sake of keeping the ability to compute
+    // these values lazily
+    /** Returns the maximum speed of the envelope */
+    /** The highest speed */
+    @JvmField val maxSpeed: Double
+
+    /** Returns the minimum speed of the envelope */
+    /** The smallest speed */
+    @JvmField val minSpeed: Double
+
+    /**
+     * The time from the start of the envelope to envelope part transitions, in microseconds. Only
+     * read using getTotalTimes.
+     */
+    private var cumulativeTimesCache: LongArray? = null
+
+    // endregion
+    // region CONSTRUCTORS
+    init {
+        assert(parts.isNotEmpty())
+
+        // check for space and speed continuity. space continuity is mandatory, speed continuity is
+        // not
+        var continuous = true
+        for (i in 0 until parts.size - 1) {
+            if (parts[i].endPos != parts[i + 1].beginPos)
+                throw OSRDError(ErrorType.EnvelopePartsNotContiguous)
+            if (!TrainPhysicsIntegrator.areSpeedsEqual(parts[i].endSpeed, parts[i + 1].beginSpeed))
+                continuous = false
+        }
+
+        // find the minimum and maximum speeds for all envelope parts
+        var minSpeed = Double.POSITIVE_INFINITY
+        var maxSpeed = Double.NEGATIVE_INFINITY
+        for (part in parts) {
+            val partMinSpeed = part.minSpeed
+            if (partMinSpeed < minSpeed) minSpeed = partMinSpeed
+            val partMaxSpeed = part.maxSpeed
+            if (partMaxSpeed > maxSpeed) maxSpeed = partMaxSpeed
+        }
+
+        // fill the part transition positions cache
+        val partPositions = DoubleArray(parts.size + 1)
+        partPositions[0] = parts[0].beginPos
+        for (i in parts.indices) partPositions[i + 1] = parts[i].endPos
+        this.partPositions = partPositions
+
+        this.parts = parts
+        this.continuous = continuous
+        this.minSpeed = minSpeed
+        this.maxSpeed = maxSpeed
+    }
+
+    // endregion
+    // region GETTERS
+    fun size(): Int {
+        return parts.size
+    }
+
+    fun get(i: Int): EnvelopePart {
+        return parts[i]
+    }
+
+    override fun getBeginPos(): Double {
+        return parts[0].beginPos
+    }
+
+    override fun getEndPos(): Double {
+        return parts[parts.size - 1].endPos
```

Comment:
```suggestion
        return parts.last().endPos
```

Very nitpick.

Okay I should probably zoom out and stop with such extreme details.